src/bt-core-io.c

```diff
@@ -480,16 +483,18 @@ struct alarm_state {
 static enum ioresult
 alarm_event_in(int fd, void* data)
 {
-  uint64_t num_expirations;
-  ssize_t res;
+  struct wake_alarm_param* param = data;
+
+  assert(param);
 
-  struct alarm_state* alarm_state = data;
-  assert(alarm_state);
-  if (alarm_state->cb)
-    alarm_state->cb(alarm_state->data);
+  if (param->cb)
+    param->cb(param->data);
+
+  remove_fd_from_epoll_loop(fd);
+  if (TEMP_FAILURE_RETRY(close(fd)) < 0)
+    ALOGW_ERRNO("close");
+  free(param);
 
-  remove_timer(fd);
-  free(data);
   return IO_OK;
```

Comment:
You need to return IO_POLL here, because you removed and closed the file descriptor. The I/O loop needs to update it's state.